### ISSUE TITLE
`java-test-fixtures` removes `main` source set from classpath, resulting in tests not finding main classes anymore

### ISSUE THREAD

**Issue by hakanai** [gradle/gradle#19946]
I had originally posted this as [a question on the forum](https://discuss.gradle.org/t/added-java-test-fixtures-plugin-now-tests-cant-find-sources/42226).

The following code in the `java-test-fixtures` plugin seems to deliberately stomp the test classpath down to including just the test code and not the main classes:

https://github.com/gradle/gradle/blob/f0d9291c04b90b59445041eaa75b2ee744162586/subprojects/plugins/src/main/java/org/gradle/api/plugins/JavaTestFixturesPlugin.java#L74-L78

However in our case, compile configurations have `transitive = false`, so we never had duplicate main classes in the first place, and this change removes from the classpath the only copy of the main classes we did have.

Removing duplicate classpath entries does seem like a fair thing to do though. Maybe there is a more reliable way to go about it?

**Comment by hakanai** [comment]
The workaround, I assume, is adding the main classes back somehow.

**Comment by hakanai** [comment]
The test sources should depend on the main sources too, but this plugin removes that for some reason.

Compile configurations not taking transitive dependencies is IMO perfectly correct behaviour, and the only reason we have to set that in our build is because Gradle behaves incorrectly in that regard in the first place.

**Comment by ljacomet** [comment]
> Compile configurations not taking transitive dependencies is IMO perfectly correct behaviour, and the only reason we have to set that in our build is because Gradle behaves incorrectly in that regard in the first place.

Can you elaborate on this?

**Comment by hakanai** [comment]
Yeah, the basic assertion: if module A depends on module B, and module B happens to depend on on module C, then while compiling module A, module C should not be on the classpath.

Including it on the classpath causes a number of issues:
- People will accidentally use code from module C
- Changes to what module B depends on will break code in module A even if their API doesn't change
- More modules are on the classpath than what is actually needed, which slows compile times

Now, Gradle's separation of 'api' and 'implementation' dependencies theoretically solves this, but either we have to wait for 100% of the external libraries we depend on to publish better metadata (which is happening, just slowly), _or_ we have to add thousands of lines to our build to adjust those modules' dependencies.

Another potential solution would be that gradle-lint-plugin fix all the issues which were breaking it in our project. Then that plugin could be used to detect modules being used in code which are not declared as dependencies. People would still see the wrong classpath when working in an IDE, but the checks would fail if they tried using those classes.

**Comment by blagerweij** [comment]
I also ran into the same issue when using java-test-fixtures and jar.enabled = false. The following buildscript fails to build:

```
plugins {
    id "java"
    id "java-test-fixtures"
}

group = 'com.example'
version = '0.0.1-SNAPSHOT'
sourceCompatibility = '17'

repositories {
	mavenCentral()
}

dependencies {
}

jar {
    enabled = false
}
```
See https://github.com/blagerweij/gradle-java-test-fixtures-bugreport

Is there a workaround? I need to disable the jar file, since our project uses Liquibase, and it's complaining about duplicate resources (once in the JAR file, once on the build/classes folder)

**Comment by luiznaac** [comment]
@blagerweij did you find any workaround on that? I'm facing the same issue right now

**Comment by nbrugger-tgm** [comment]
Please re-open this issue.
Currently, it is not possible (as @blagerweij laid out) to have the jar task disabled and use the classfiles instead which causes issues with plugins that require test to be independent of jar or just disabling the jar.

Adding this to the class task does **not** help (i dont know why):
```kotlin
 this.classpath.filter {
        if(it.absolutePath.equals(tasks.jar.get().outputs.files.singleFile.absolutePath))
            logger.lifecycle("removed from test classpath = ${ it.absolutePath }")
        !it.absolutePath.equals(tasks.jar.get().outputs.files.singleFile.absolutePath)
    }.plus(sourceSets.main.map { it.output.singleFile })
    ```
    or
    ```kotlin
     this.classpath = this.classpath.filter {
        if(it.absolutePath.equals(tasks.jar.get().outputs.files.singleFile.absolutePath))
            logger.lifecycle("removed from test classpath = ${ it.absolutePath }")
        !it.absolutePath.equals(tasks.jar.get().outputs.files.singleFile.absolutePath)
    }.plus(sourceSets.main.map { it.output.singleFile })
    ```

**Comment by nbrugger-tgm** [comment]
@luiznaac 
Here is a solution:
```kotlin
// The `java-test-fixtures` plugin breaks test-classpath so this fixes it by resetting it.
// https://github.com/gradle/gradle/issues/19946
configurations.testRuntimeClasspath {
    exclude(module = project.name) //This excludes the JAR
}

tasks.test {
    classpath += files(sourceSets.main.map { it.output }) //This includes the class-files
}
```
